aten/src/ATen/native/quantized/cpu/q_avgpool.cpp

```diff
@@ -12,212 +14,265 @@ namespace at {
-    Tensor& output,
+template <typename scalar_t>
+Tensor q_avg_pool2d(
     const Tensor& input,
-    IntArrayRef output_size) {
-  const auto output_shape = get_output_shape(input, output_size);
-  TORCH_CHECK(
-      output.is_quantized() && output.sizes() == output_shape,
-      "Output Tensor must be quantized and have a shape of ",
-      "{ ",
-      output_shape,
-      " }.");
-  adaptive_avg_pool2d_out_template(output, input, output_shape);
-  return output;
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    bool ceil_mode,
+    bool count_include_pad,
+    c10::optional<int64_t> divisor_override) {
+
+    int kW, kH, dW, dH, padW, padH;
+    std::tie(kW, kH) = get_kernel(kernel_size);
+    std::tie(dW, dH) = get_stride(stride, kW, kH);
+    std::tie(padW, padH) = get_padding(padding);
+
+    const int64_t nbatch = input.ndimension() == 4 ? input.size(-4) : 1;
+    const int64_t nInputPlane = input.size(-3);
+    const int64_t inputHeight = input.size(-2);
+    const int64_t inputWidth = input.size(-1);
+
+    TORCH_CHECK(!divisor_override.has_value() || divisor_override.value() != 0,
+      "divisor must be not zero");
+
+    auto output_shape = get_output_shape(input, kW, kH, dW, dH, padW, padH, ceil_mode);
+    const int64_t outputHeight = output_shape[output_shape.size() - 2];
+    const int64_t outputWidth = output_shape[output_shape.size() - 1];
+    if (input.is_contiguous(c10::MemoryFormat::ChannelsLast)) {
+      auto output = at::_empty_affine_quantized(
+          output_shape,
+          input.options(),
+          input.q_scale(),
+          input.q_zero_point(),
+          input.suggest_memory_format());
+      // fast path for channel last: qavg_pool_2d_nhwc_stub
+      if (output_shape.size() == 3) {
+        qavg_pool2d_nhwc_stub(
+            input.device().type(),
+            input,
+            output,
+            0,
+            nInputPlane,
+            inputWidth,
+            inputHeight,
+            outputWidth,
+            outputHeight,
+            kW,
+            kH,
+            dW,
+            dH,
+            padW,
+            padH,
+            count_include_pad,
+            divisor_override);
+      } else {
+        at::parallel_for(0, nbatch, 0, [&](int64_t start, int64_t end) {
+          for (auto b = start; b < end; b++) {
+            qavg_pool2d_nhwc_stub(
+                input.device().type(),
+                input,
+                output,
+                b,
+                nInputPlane,
+                inputWidth,
+                inputHeight,
+                outputWidth,
+                outputHeight,
+                kW,
+                kH,
+                dW,
+                dH,
+                padW,
+                padH,
+                count_include_pad,
+                divisor_override);
+          }
+        });
+      }
+      return output;
+    } else {
+      auto output = at::_empty_affine_quantized(
```

Comment:
Are you expecting input to be contiguous here, or `avg_pool2d_out_frame` can work with any strides?